<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>QuickStart &mdash; jGEM 0.9.0 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.9.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="jGEM 0.9.0 documentation" href="index.html" />
    <link rel="next" title="jGEM packages" href="modules.html" />
    <link rel="prev" title="Installation" href="install.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput,
div.nbinput > div,
div.nbinput div[class^=highlight],
div.nbinput div[class^=highlight] pre,
div.nboutput,
div.nboutput > div,
div.nboutput div[class^=highlight],
div.nboutput div[class^=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput div[class^=highlight] pre {
    line-height: unset;
}

/* input/output containers */
div.nbinput,
div.nboutput {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
}

/* input container */
div.nbinput {
    padding-top: 5px;
}

/* last container */
div.nblast {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput > :first-child pre {
    color: #303F9F;
}

/* output prompt */
div.nboutput > :first-child pre {
    color: #D84315;
}

/* all prompts */
div.nbinput > :first-child[class^=highlight],
div.nboutput > :first-child[class^=highlight],
div.nboutput > :first-child {
    min-width: 11ex;
    padding-top: 0.4em;
    padding-right: 0.4em;
    text-align: right;
    flex: 0;
}

/* input/output area */
div.nbinput > :nth-child(2)[class^=highlight],
div.nboutput > :nth-child(2),
div.nboutput > :nth-child(2)[class^=highlight] {
    padding: 0.4em;
    -webkit-flex: 1;
    flex: 1;
}

/* input area */
div.nbinput > :nth-child(2)[class^=highlight] {
    border: 1px solid #cfcfcf;
    border-radius: 2px;
    background: #f7f7f7;
}

/* override MathJax center alignment in output cells */
div.nboutput div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.pngmath center alignment in output cells */
div.nboutput div.math p {
    text-align: left;
}

/* standard error */
div.nboutput  > :nth-child(2).stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-bold { font-weight: bold; }
</style>
<div class="section" id="QuickStart">
<h1>QuickStart<a class="headerlink" href="#QuickStart" title="Permalink to this headline">¶</a></h1>
<p>We will do the assembly and merging using the data in the tests/data
directory in the following order:</p>
<ol class="arabic simple">
<li>Preparing sample info spreadsheet</li>
<li>Preparing inputs</li>
<li>Assembling individual samples</li>
<li>Evalulation of individual assemblies</li>
<li>Preparing for merge</li>
<li>Merging</li>
<li>Evaluation of the merged assembly</li>
</ol>
<p>The example data is the chromosome 1 portion of the RNASeq data from the
neurons in superior central nucleus raphe, medial part (CSm) and dosal
nucleus raphe (DR) in Fev-Cre mouse line which labels serotoninergic
neurons. (Fev-Cre is also called ePet-Cre.)</p>
<p>First we need to change logging level for jupyter notebook and activate
inline matplotlib figure:</p>
<div class="nbinput nblast container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [89]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># This is to change logging level of jupyter notebook</span>
<span class="kn">from</span> <span class="nn">importlib</span> <span class="kn">import</span> <span class="nb">reload</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="nb">reload</span><span class="p">(</span><span class="n">logging</span><span class="p">)</span>
<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%(levelname)s</span><span class="s1">:</span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span> <span class="n">datefmt</span><span class="o">=</span><span class="s1">&#39;%I:%M:%S&#39;</span><span class="p">)</span>

<span class="c1"># This is to show matplotlib output in the notebook</span>
<span class="o">%</span><span class="k">matplotlib</span> inline
</pre></div>
</div>
</div>
<div class="section" id="Sample-Info">
<h2>Sample Info<a class="headerlink" href="#Sample-Info" title="Permalink to this headline">¶</a></h2>
<p>For this quickstart, we will use the Excel file <strong>sampleinfo.xlsx</strong> in
the tests/data directory. (In general, it is useful to have an Excel
spreadsheet or a tab separated text file which contains following fields
for each sample.)</p>
<ul class="simple">
<li><strong>name</strong>: unique sample name</li>
<li><strong>bwfile</strong>: location of normalized coverage in BIGWIG format</li>
<li><strong>sjfile</strong>: location of splice junction table</li>
<li><strong>sample_id</strong>: unique sample identifying number</li>
<li><strong>group</strong>: sample group</li>
<li><strong>bamfile</strong>: BAM file location</li>
<li><strong>sjtabfile</strong>: location of STAR SJ.out.tab</li>
<li><strong>aligned</strong>: number of aligned reads</li>
</ul>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [6]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">PD</span>

<span class="n">si</span> <span class="o">=</span> <span class="n">PD</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s1">&#39;../tests/data/sampleinfo.xlsx&#39;</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">si</span><span class="p">[[</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;bam&#39;</span><span class="p">,</span><span class="s1">&#39;bigwig&#39;</span><span class="p">,</span><span class="s1">&#39;sjtab&#39;</span><span class="p">,</span><span class="s1">&#39;aligned&#39;</span><span class="p">]])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<div class="highlight"><pre>
               name                   bam                    bigwig  \
0   Fev_DR_m70_1623   Fev_DR_m70_1623.bam   Fev_DR_m70_1623.chr1.bw
1  Fev_CSm_m70_1624  Fev_CSm_m70_1624.bam  Fev_CSm_m70_1624.chr1.bw
2   Fev_DR_m71_1625   Fev_DR_m71_1625.bam   Fev_DR_m71_1625.chr1.bw
3  Fev_CSm_m71_1626  Fev_CSm_m71_1626.bam  Fev_CSm_m71_1626.chr1.bw

                              sjtab   aligned
0   Fev_DR_m70_1623.chr1.SJ.out.tab  24313418
1  Fev_CSm_m70_1624.chr1.SJ.out.tab  25857770
2   Fev_DR_m71_1625.chr1.SJ.out.tab  26260366
3  Fev_CSm_m71_1626.chr1.SJ.out.tab  26083777
</pre></div></div>
</div>
<p>We&#8217;ll adjust actual locations of bigwig and SJ.out.tab files:</p>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [7]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>

<span class="n">BASEDIR</span> <span class="o">=</span> <span class="s1">&#39;../tests/data&#39;</span>
<span class="n">si</span><span class="p">[</span><span class="s1">&#39;bwfile&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">BASEDIR</span><span class="p">,</span> <span class="s2">&quot;bigwig&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">si</span><span class="p">[</span><span class="s1">&#39;bigwig&#39;</span><span class="p">]]</span>
<span class="n">si</span><span class="p">[</span><span class="s1">&#39;sjtabfile&#39;</span><span class="p">]</span> <span class="o">=</span>  <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">BASEDIR</span><span class="p">,</span> <span class="s2">&quot;SJ&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">si</span><span class="p">[</span><span class="s1">&#39;sjtab&#39;</span><span class="p">]]</span>

<span class="k">print</span><span class="p">(</span><span class="n">si</span><span class="p">[[</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;bwfile&#39;</span><span class="p">,</span><span class="s1">&#39;sjtabfile&#39;</span><span class="p">]])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<div class="highlight"><pre>
               name                                         bwfile  \
0   Fev_DR_m70_1623   ../tests/data/bigwig/Fev_DR_m70_1623.chr1.bw
1  Fev_CSm_m70_1624  ../tests/data/bigwig/Fev_CSm_m70_1624.chr1.bw
2   Fev_DR_m71_1625   ../tests/data/bigwig/Fev_DR_m71_1625.chr1.bw
3  Fev_CSm_m71_1626  ../tests/data/bigwig/Fev_CSm_m71_1626.chr1.bw

                                           sjtabfile
0   ../tests/data/SJ/Fev_DR_m70_1623.chr1.SJ.out.tab
1  ../tests/data/SJ/Fev_CSm_m70_1624.chr1.SJ.out.tab
2   ../tests/data/SJ/Fev_DR_m71_1625.chr1.SJ.out.tab
3  ../tests/data/SJ/Fev_CSm_m71_1626.chr1.SJ.out.tab
</pre></div></div>
</div>
</div>
<div class="section" id="Preparing-Inputs">
<h2>Preparing Inputs<a class="headerlink" href="#Preparing-Inputs" title="Permalink to this headline">¶</a></h2>
<p>You can prepare your normalized coverage bigwig files and junction
tables (format will be explained below) anyway you want, but here is one
way you can prepare these from the outpus of STAR mapper (BAM and
SJ.out.tab).</p>
<p>To generate coverage bigwigs, use function
<a class="reference external" href="modules.html#jgem.bigwig.bam2bw">jgem.bigwig.bam2bw</a>:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">jgem</span> <span class="kn">import</span> <span class="n">bigwig</span> <span class="k">as</span> <span class="n">BW</span>
<span class="kn">from</span> <span class="nn">jgem</span> <span class="kn">import</span> <span class="n">utils</span> <span class="k">as</span> <span class="n">UT</span>

<span class="n">chromsizes</span> <span class="o">=</span> <span class="n">UT</span><span class="o">.</span><span class="n">chromsizes</span><span class="p">(</span><span class="s1">&#39;mm10&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">bam</span><span class="p">,</span> <span class="n">bw</span><span class="p">,</span> <span class="n">aligned</span> <span class="ow">in</span> <span class="n">si</span><span class="p">[[</span><span class="s1">&#39;bamfile&#39;</span><span class="p">,</span><span class="s1">&#39;bwfile&#39;</span><span class="p">,</span><span class="s1">&#39;aligned&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
    <span class="n">BW</span><span class="o">.</span><span class="n">bam2bw</span><span class="p">(</span><span class="n">bam</span><span class="p">,</span><span class="n">chromsizes</span><span class="p">,</span><span class="n">bw</span><span class="p">,</span><span class="n">aligned</span><span class="p">)</span>
</pre></div>
</div>
<p>To generate junction files from SJ.out.tab, use function
<a class="reference external" href="modules.html#jgem.gtfgffbed.sjtab2sjbed">jgem.gtfgffbed.sjtab2sjbed</a>:</p>
<p>First create destinations:</p>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [8]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="n">si</span><span class="p">[</span><span class="s1">&#39;sjfile&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">BASEDIR</span><span class="p">,</span> <span class="s2">&quot;SJ&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">[:</span><span class="o">-</span><span class="mi">10</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;sj.bed.gz&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">si</span><span class="p">[</span><span class="s1">&#39;sjtab&#39;</span><span class="p">]]</span>
<span class="k">print</span><span class="p">(</span><span class="n">si</span><span class="p">[[</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;sjfile&#39;</span><span class="p">]])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<div class="highlight"><pre>
               name                                            sjfile
0   Fev_DR_m70_1623   ../tests/data/SJ/Fev_DR_m70_1623.chr1.sj.bed.gz
1  Fev_CSm_m70_1624  ../tests/data/SJ/Fev_CSm_m70_1624.chr1.sj.bed.gz
2   Fev_DR_m71_1625   ../tests/data/SJ/Fev_DR_m71_1625.chr1.sj.bed.gz
3  Fev_CSm_m71_1626  ../tests/data/SJ/Fev_CSm_m71_1626.chr1.sj.bed.gz
</pre></div></div>
</div>
<p>Then, convert from SJ.out.tab to appropriate BED file.</p>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [4]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="o">%%</span><span class="k">time</span>
from jgem import gtfgffbed as GGB

for sjtab, sjbed, aligned in si[[&#39;sjtabfile&#39;,&#39;sjfile&#39;,&#39;aligned&#39;]].values:
    GGB.sjtab2sjbed(sjtab,sjbed,aligned)
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<div class="highlight"><pre>
CPU times: user 376 ms, sys: 33.1 ms, total: 409 ms
Wall time: 513 ms
</pre></div></div>
</div>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [7]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="n">ls</span>  <span class="o">../</span><span class="n">tests</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">SJ</span><span class="o">/*.</span><span class="n">chr1</span><span class="o">.</span><span class="n">sj</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<div class="highlight"><pre>
../tests/data/SJ/Fev_CSm_m70_1624.chr1.sj.bed.gz
../tests/data/SJ/Fev_CSm_m71_1626.chr1.sj.bed.gz
../tests/data/SJ/Fev_DR_m70_1623.chr1.sj.bed.gz
../tests/data/SJ/Fev_DR_m71_1625.chr1.sj.bed.gz
</pre></div></div>
</div>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [9]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># check the contents</span>
<span class="kn">from</span> <span class="nn">jgem</span> <span class="kn">import</span> <span class="n">utils</span> <span class="k">as</span> <span class="n">UT</span>

<span class="n">sj</span> <span class="o">=</span> <span class="n">UT</span><span class="o">.</span><span class="n">read_pandas</span><span class="p">(</span><span class="n">sjbed</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">,</span><span class="s1">&#39;st&#39;</span><span class="p">,</span><span class="s1">&#39;ed&#39;</span><span class="p">,</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;ucnt&#39;</span><span class="p">,</span><span class="s1">&#39;strand&#39;</span><span class="p">,</span><span class="s1">&#39;mcnt&#39;</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="n">sj</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<div class="highlight"><pre>
    chr       st       ed                 name      ucnt strand      mcnt
0  chr1  3121707  3142317   CT/AC-k0-u4-m0-o46  0.153352      -  0.000000
1  chr1  3207318  3213438  CT/AC-k0-u10-m0-o47  0.383380      -  0.000000
2  chr1  3271573  3302745   GT/AG-k0-u0-m1-o37  0.000000      +  0.038338
3  chr1  3421902  3670551  CT/AC-k1-u20-m0-o43  0.766760      -  0.000000
4  chr1  3660890  3670551   CT/AC-k0-u2-m0-o41  0.076676      -  0.000000
</pre></div></div>
</div>
<div class="section" id="SJ-txt-file-format">
<h3>SJ txt file format<a class="headerlink" href="#SJ-txt-file-format" title="Permalink to this headline">¶</a></h3>
<p>As shown above, the tab separated junction file should contain:</p>
<ul class="simple">
<li><strong>chr</strong>: chromosome name</li>
<li><strong>st</strong>: start position</li>
<li><strong>ed</strong>: end position</li>
<li><strong>name</strong>: junction name</li>
<li><strong>strand</strong>: strand of junction</li>
<li><strong>ucnt</strong>: unique reads (normalized to 1M mapped reads)</li>
<li><strong>mcnt</strong>: non-unique (multimapping) reads (normalized)</li>
</ul>
</div>
</div>
<div class="section" id="Assembling-individual-samples">
<h2>Assembling individual samples<a class="headerlink" href="#Assembling-individual-samples" title="Permalink to this headline">¶</a></h2>
<p>Use <a class="reference external" href="modules.html#jgem.filenames.FileNames">FileNames class</a> to
specify inputs (bigwig file and sjbed file) and output directory, then
use <a class="reference external" href="modules.html#jgem.assembler.Assembler">Assembler class</a> to do
the assembly.</p>
<div class="nbinput nblast container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [13]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">jgem</span> <span class="kn">import</span> <span class="n">filenames</span> <span class="k">as</span> <span class="n">FN</span>
<span class="kn">from</span> <span class="nn">jgem</span> <span class="kn">import</span> <span class="n">assembler</span> <span class="k">as</span> <span class="n">AS</span>
</pre></div>
</div>
</div>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [14]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="o">%%</span><span class="k">time</span>
# assemble all samples
for name,bwfile,sjfile in si[[&#39;name&#39;,&#39;bwfile&#39;,&#39;sjfile&#39;]].values:
    fn = FN.FileNames(name, bwfile, sjfile, outdir=&#39;../out&#39;)
    p = AS.Assembler(fn,saveintermediates=False)
    p.assemble()
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<div class="highlight"><pre>
CPU times: user 2min 14s, sys: 9.67 s, total: 2min 23s
Wall time: 4min 51s
</pre></div></div>
</div>
<p>In the output directory, several files are generated:</p>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [4]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="n">ls</span> <span class="o">../</span><span class="n">out</span><span class="o">/</span><span class="n">Fev_DR_m70_1623</span><span class="o">*</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<div class="highlight"><pre>
../out/Fev_DR_m70_1623.assemble.params.txt.gz
../out/Fev_DR_m70_1623.assemble.stats.txt.gz
../out/Fev_DR_m70_1623.ex.bed.gz
../out/Fev_DR_m70_1623.ex.txt.gz
../out/Fev_DR_m70_1623.findsecovth.params.txt.gz
../out/Fev_DR_m70_1623.findsecovth.pdf
../out/Fev_DR_m70_1623.genes.bed.gz
../out/Fev_DR_m70_1623.sj.bed.gz
../out/Fev_DR_m70_1623.sj.txt.gz
</pre></div></div>
</div>
<p>Main outputs are .ex.txt.gz, .sj.txt.gz. These contain exons and
junctions in tab separated format. BED files are provided for viewing
elements (genes,exons,junctions) with browsers like IGV
(<a class="reference external" href="https://www.broadinstitute.org/igv/">https://www.broadinstitute.org/igv/</a>). The screenshot below is a segment
of chromosome 1 showing genes.bed files from the 4 samples.</p>
<div class="figure" id="id1">
<img alt="IGV2" src="_images/WMSUAE41D6ONM36O9K5LYVIFO0T7EKLM.png" />
<p class="caption"><span class="caption-text">IGV2</span></p>
</div>
<p>Other files (assemble.params.txt.gz, assemble.stats.txt.gz,
findsecovth.params.txt.gz, findsecovth.pdf) contain parameters and
statistics of the assembly.</p>
</div>
<div class="section" id="Evaluation-of-individual-assemblies">
<h2>Evaluation of individual assemblies<a class="headerlink" href="#Evaluation-of-individual-assemblies" title="Permalink to this headline">¶</a></h2>
<p>We will evaluate the assemblies against Gencode.VM4 annotation
(<a class="reference external" href="http://www.gencodegenes.org/mouse_releases/4.html">http://www.gencodegenes.org/mouse_releases/4.html</a>). For this, we use
<a class="reference external" href="modules.html#jgem.evalutate.EvalNames">EvalNames class</a> and
<a class="reference external" href="modules.html#jgem.evalutate.EvalMatch">EvalMatch class</a>.</p>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [94]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="o">%%</span><span class="k">time</span>
from jgem import evaluate as EV

ens = {} # keep EvalNames objects
ems = {} # keep EvalMatch object

# Gencode.VM4
en_gen4 = EV.EvalNames(&#39;../tests/data/assemblies/gencode.vM4.chr1&#39;, code=&#39;gen4&#39;, outdir=&#39;../out/&#39;)

# For each sample
for sid, name, bwfile, sjfile in si[[&#39;sample_id&#39;, &#39;name&#39;,&#39;bwfile&#39;,&#39;sjfile&#39;]].values:
    print(&#39;process {0}...&#39;.format(name))
    ens[name] = EV.EvalNames(&#39;../out/&#39;+name,  code=&#39;a&#39;+str(sid), outdir=&#39;../out/&#39;)
    # code indicates assembly
    ems[name] = EV.EvalMatch(en_gen4, ens[name], bwfile, sjfile, datacode=&#39;d&#39;+str(sid), binsize=100)
    # datacode indicate data (bwfile,sjfile)
    ems[name].calculate()
</pre></div>
</div>
</div>
<div class="nboutput container">
<div class="container">
</div>
<div class="container">
<div class="highlight"><pre>
process Fev_DR_m70_1623...
</pre></div></div>
</div>
<div class="nboutput container">
<div class="container">
</div>
<div class="stderr container">
<div class="highlight"><pre>
INFO:[i] detected1:7862,        matched:5584,   (detected2:5834),       ratio:0.71,     (ratio2:0.96)
INFO:[5] detected1:2336,        matched:1086,   (detected2:1218),       ratio:0.46,     (ratio2:0.89)
INFO:[3] detected1:2632,        matched:1076,   (detected2:1311),       ratio:0.41,     (ratio2:0.82)
INFO:[s] detected1:944, matched:97,     (detected2:400),        ratio:0.10,     (ratio2:0.24)
INFO:[j] detected1:6892,        matched:6892,   (detected2:7636),       ratio:1.00,     (ratio2:0.90)
INFO:[5b] detected1:2336,       matched:1566,   (detected2:8763),       ratio:0.67,     (ratio2:0.18)
INFO:[3b] detected1:2632,       matched:1708,   (detected2:8763),       ratio:0.65,     (ratio2:0.19)
INFO:[sb] detected1:944,        matched:263,    (detected2:8763),       ratio:0.28,     (ratio2:0.03)
</pre></div></div>
</div>
<div class="nboutput container">
<div class="container">
</div>
<div class="container">
<div class="highlight"><pre>
process Fev_CSm_m70_1624...
</pre></div></div>
</div>
<div class="nboutput container">
<div class="container">
</div>
<div class="stderr container">
<div class="highlight"><pre>
INFO:[i] detected1:8034,        matched:5858,   (detected2:6191),       ratio:0.73,     (ratio2:0.95)
INFO:[5] detected1:2436,        matched:1130,   (detected2:1207),       ratio:0.46,     (ratio2:0.94)
INFO:[3] detected1:2742,        matched:1137,   (detected2:1293),       ratio:0.41,     (ratio2:0.88)
INFO:[s] detected1:949, matched:110,    (detected2:477),        ratio:0.12,     (ratio2:0.23)
INFO:[j] detected1:7153,        matched:7151,   (detected2:8118),       ratio:1.00,     (ratio2:0.88)
INFO:[5b] detected1:2436,       matched:1637,   (detected2:9168),       ratio:0.67,     (ratio2:0.18)
INFO:[3b] detected1:2742,       matched:1820,   (detected2:9168),       ratio:0.66,     (ratio2:0.20)
INFO:[sb] detected1:949,        matched:284,    (detected2:9168),       ratio:0.30,     (ratio2:0.03)
</pre></div></div>
</div>
<div class="nboutput container">
<div class="container">
</div>
<div class="container">
<div class="highlight"><pre>
process Fev_DR_m71_1625...
</pre></div></div>
</div>
<div class="nboutput container">
<div class="container">
</div>
<div class="stderr container">
<div class="highlight"><pre>
INFO:[i] detected1:8262,        matched:5950,   (detected2:6375),       ratio:0.72,     (ratio2:0.93)
INFO:[5] detected1:2517,        matched:1148,   (detected2:1275),       ratio:0.46,     (ratio2:0.90)
INFO:[3] detected1:2865,        matched:1175,   (detected2:1387),       ratio:0.41,     (ratio2:0.85)
INFO:[s] detected1:1014,        matched:98,     (detected2:400),        ratio:0.10,     (ratio2:0.24)
INFO:[j] detected1:7336,        matched:7334,   (detected2:8416),       ratio:1.00,     (ratio2:0.87)
INFO:[5b] detected1:2517,       matched:1678,   (detected2:9437),       ratio:0.67,     (ratio2:0.18)
INFO:[3b] detected1:2865,       matched:1886,   (detected2:9437),       ratio:0.66,     (ratio2:0.20)
INFO:[sb] detected1:1014,       matched:281,    (detected2:9437),       ratio:0.28,     (ratio2:0.03)
</pre></div></div>
</div>
<div class="nboutput container">
<div class="container">
</div>
<div class="container">
<div class="highlight"><pre>
process Fev_CSm_m71_1626...
</pre></div></div>
</div>
<div class="nboutput container">
<div class="container">
</div>
<div class="stderr container">
<div class="highlight"><pre>
INFO:[i] detected1:8003,        matched:5726,   (detected2:6048),       ratio:0.72,     (ratio2:0.95)
INFO:[5] detected1:2456,        matched:1169,   (detected2:1263),       ratio:0.48,     (ratio2:0.93)
INFO:[3] detected1:2746,        matched:1118,   (detected2:1324),       ratio:0.41,     (ratio2:0.84)
INFO:[s] detected1:969, matched:92,     (detected2:408),        ratio:0.09,     (ratio2:0.23)
INFO:[j] detected1:7085,        matched:7083,   (detected2:7989),       ratio:1.00,     (ratio2:0.89)
INFO:[5b] detected1:2456,       matched:1681,   (detected2:9043),       ratio:0.68,     (ratio2:0.19)
INFO:[3b] detected1:2746,       matched:1803,   (detected2:9043),       ratio:0.66,     (ratio2:0.20)
INFO:[sb] detected1:969,        matched:261,    (detected2:9043),       ratio:0.27,     (ratio2:0.03)
</pre></div></div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<div class="highlight"><pre>
CPU times: user 23.2 s, sys: 1.58 s, total: 24.8 s
Wall time: 30.2 s
</pre></div></div>
</div>
<p>INFO outputs indicates counts of gencode exons/junctions with positive
coverage (detected1), matched exons/junctions (matched) and counts of
assembly (detected2) and ratios (matched/detected1, matched/detected2).</p>
<p>These are assesed within different exon classes: <strong>i</strong>, internal exon
(exons bounded by splice junctions); <strong>5</strong>, 5&#8217;exons; <strong>3</strong>, 3&#8217;exons;
<strong>s</strong>, single exons; and junctions (<strong>j</strong>). Classes with <strong>b</strong>,
(<strong>5b</strong>,<strong>3b</strong>,<strong>sb</strong>) allow matching to different classes of exons.</p>
<p>Matching exons with junction boundary requires these boundaries to
match. That is, to match internal exons, they have to be exactly the
same (including strand). To match 5&#8217;exons, donor site has to the same
but the other side does not have to be. To match single exon, they just
need to have some overlap.</p>
<p>So, we have almost perfect detection of junctions. About 70% of internal
exons are detected. 5&#8217;,3&#8217; exons are worse (~40-45%) but if allowed to
match to other class of exons, then detection rate get close to internal
exon (~65-67%). Single exon detection rates are worst and about 10%.</p>
<p>To visualize the detection percentages:</p>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [104]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># get dataframe containing detection percentages</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">ems</span><span class="p">[</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">get_detection_percentages</span><span class="p">()</span>
<span class="c1">#print(df)</span>
<span class="n">df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>Out[104]:
</pre></div>
</div>
<div class="container">
<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>%detected 1</th>
      <th>%detected 2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>i</th>
      <td>71.548169</td>
      <td>94.675926</td>
    </tr>
    <tr>
      <th>5</th>
      <td>47.597720</td>
      <td>92.557403</td>
    </tr>
    <tr>
      <th>5b</th>
      <td>68.444625</td>
      <td>18.588964</td>
    </tr>
    <tr>
      <th>3</th>
      <td>40.713765</td>
      <td>84.441088</td>
    </tr>
    <tr>
      <th>3b</th>
      <td>65.659141</td>
      <td>19.938074</td>
    </tr>
    <tr>
      <th>s</th>
      <td>9.494324</td>
      <td>22.549020</td>
    </tr>
    <tr>
      <th>sb</th>
      <td>26.934985</td>
      <td>2.886210</td>
    </tr>
    <tr>
      <th>j</th>
      <td>99.971771</td>
      <td>88.659407</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [120]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="c1"># plot all</span>
<span class="kn">import</span> <span class="nn">matplotlib.pylab</span> <span class="kn">as</span> <span class="nn">P</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axr</span> <span class="o">=</span> <span class="n">P</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">ems</span><span class="p">),</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">ems</span><span class="p">),</span><span class="mi">3</span><span class="p">))</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ems</span><span class="p">):</span>
    <span class="n">em</span> <span class="o">=</span> <span class="n">ems</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">em</span><span class="o">.</span><span class="n">get_detection_percentages</span><span class="p">()</span>
    <span class="n">df</span><span class="o">.</span><span class="n">ix</span><span class="p">[[</span><span class="s1">&#39;i&#39;</span><span class="p">,</span><span class="s1">&#39;5&#39;</span><span class="p">,</span><span class="s1">&#39;3&#39;</span><span class="p">,</span><span class="s1">&#39;s&#39;</span><span class="p">,</span><span class="s1">&#39;j&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s1">&#39;bar&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axr</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">legend</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
    <span class="n">axr</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<img alt="_images/quickstart_29_0.png" src="_images/quickstart_29_0.png" />
</div>
</div>
<p>We can also plot the detection sensitivity against coverages. We expect
detection rate is better at bigger coverage.</p>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [44]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ro-&#39;</span><span class="p">,</span><span class="s1">&#39;go-&#39;</span><span class="p">,</span><span class="s1">&#39;bo-&#39;</span><span class="p">,</span><span class="s1">&#39;mo-&#39;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ems</span><span class="p">):</span>
    <span class="n">em</span> <span class="o">=</span> <span class="n">ems</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">i</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
        <span class="n">axr</span> <span class="o">=</span> <span class="n">em</span><span class="o">.</span><span class="n">plot_sensitivity</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">axr</span> <span class="o">=</span> <span class="n">em</span><span class="o">.</span><span class="n">plot_sensitivity</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">axr</span><span class="o">=</span><span class="n">axr</span><span class="p">,</span> <span class="n">ypos</span><span class="o">=</span><span class="n">i</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<img alt="_images/quickstart_31_0.png" src="_images/quickstart_31_0.png" />
</div>
</div>
<p>These plots show percent of detected exons calculated in bins sorted by
coverage (binsize 100). Numbers in inset indicate normalized area under
the curve (AUC, which ranges from 0 to 1). Dashed lines in 5&#8217;,3&#8217; exon
panels correspond to 5b and 3b category.</p>
<p>How about the precision of the match? We can assess this by length ratio
of matched exons:</p>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [58]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">ems</span><span class="p">:</span>
    <span class="n">ems</span><span class="p">[</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">plot_ratio</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput container">
<div class="container">
</div>
<div class="container">
<img alt="_images/quickstart_34_0.png" src="_images/quickstart_34_0.png" />
</div>
</div>
<div class="nboutput container">
<div class="container">
</div>
<div class="container">
<img alt="_images/quickstart_34_1.png" src="_images/quickstart_34_1.png" />
</div>
</div>
<div class="nboutput container">
<div class="container">
</div>
<div class="container">
<img alt="_images/quickstart_34_2.png" src="_images/quickstart_34_2.png" />
</div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<img alt="_images/quickstart_34_3.png" src="_images/quickstart_34_3.png" />
</div>
</div>
<p>Averages shown inside are geometric averages. Overall length tends to be
a bit longer.</p>
<p>For some genes, we see partial or disconnected reconstruction of the
structure. An example is shown below:</p>
<div class="figure">
<img alt="" src="_images/UCNYV53JQT9QG8GJQSS36GKMGSIT22C2.png" />
</div>
<p>These incomplete reconstructions are due to missing junctions or
insufficient exon coverages.</p>
<p>To evaluate the <em>completeness</em> of the reconstruction, we will calculate
three measures. One is the ratio of gene length. We will call this GLC
(gene length completenesss). The second measure is the ratio of exon
counts (ECC, exon count completeness), and the third is the ratio of
junction counts (JCC, junction count completeness).</p>
<p>We can visualize these measures by
<a class="reference external" href="modules.html#jgem.evaluate.EvalMatch.plot_completeness">plot_completeness</a>
method.</p>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [119]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="n">axr</span> <span class="o">=</span> <span class="n">ems</span><span class="p">[</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">plot_completeness</span><span class="p">(</span><span class="n">pw</span><span class="o">=</span><span class="s1">&#39;fdat&#39;</span><span class="p">,</span> <span class="n">ca</span><span class="o">=</span><span class="s1">&#39;g.-&#39;</span><span class="p">,</span> <span class="n">cf</span><span class="o">=</span><span class="s1">&#39;r-&#39;</span><span class="p">,</span>  <span class="n">cd</span><span class="o">=</span><span class="s1">&#39;b.&#39;</span><span class="p">,</span> <span class="n">xlim</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<img alt="_images/quickstart_39_0.png" src="_images/quickstart_39_0.png" />
</div>
</div>
<p>Blue dots are genes, green dots are binned averages, red lines are
sigmoid function fit to blue dots and red dashes are 99% position of the
sigmoidal fit.</p>
<p>Below, we&#8217;ll plot binned average of completeness measures for all 4
samples.</p>
<div class="nbinput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [131]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span><span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ro-&#39;</span><span class="p">,</span><span class="s1">&#39;go-&#39;</span><span class="p">,</span><span class="s1">&#39;bo-&#39;</span><span class="p">,</span><span class="s1">&#39;mo-&#39;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ems</span><span class="p">):</span>
    <span class="n">em</span> <span class="o">=</span> <span class="n">ems</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">i</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
        <span class="n">axr</span> <span class="o">=</span> <span class="n">em</span><span class="o">.</span><span class="n">plot_completeness</span><span class="p">(</span><span class="n">pw</span><span class="o">=</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="n">ca</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">xlim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span> <span class="n">label</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">axr</span> <span class="o">=</span> <span class="n">em</span><span class="o">.</span><span class="n">plot_completeness</span><span class="p">(</span><span class="n">pw</span><span class="o">=</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="n">ca</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">xlim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span> <span class="n">axr</span><span class="o">=</span><span class="n">axr</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">name</span><span class="p">)</span>
<span class="n">P</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>Out[131]:
</pre></div>
</div>
<div class="highlight-none"><div class="highlight"><pre>
<span></span>&lt;matplotlib.legend.Legend at 0x113205470&gt;
</pre></div>
</div>
</div>
<div class="nboutput nblast container">
<div class="container">
</div>
<div class="container">
<img alt="_images/quickstart_41_1.png" src="_images/quickstart_41_1.png" />
</div>
</div>
</div>
</div>
<div class="section" id="Preparing-for-merge">
<h1>Preparing for merge<a class="headerlink" href="#Preparing-for-merge" title="Permalink to this headline">¶</a></h1>
<p>Merging assembies is necessary to obtain a more complete picture of gene
structures. Merging should also improve the completeness of the
reconstruction.</p>
<p>For merging, we will repeat the assembly process but uses aggregated
junctions and coverage bigwig generated from all exons from all
assemblies. We need to prepare these two files first.</p>
<div class="section" id="prepare-aggregated-junction-file">
<h2>prepare aggregated junction file<a class="headerlink" href="#prepare-aggregated-junction-file" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast container">
<div class="highlight-none"><div class="highlight"><pre>
<span></span>In [ ]:
</pre></div>
</div>
<div class="highlight-ipython3"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">QuickStart</a><ul>
<li><a class="reference internal" href="#Sample-Info">Sample Info</a></li>
<li><a class="reference internal" href="#Preparing-Inputs">Preparing Inputs</a><ul>
<li><a class="reference internal" href="#SJ-txt-file-format">SJ txt file format</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Assembling-individual-samples">Assembling individual samples</a></li>
<li><a class="reference internal" href="#Evaluation-of-individual-assemblies">Evaluation of individual assemblies</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Preparing-for-merge">Preparing for merge</a><ul>
<li><a class="reference internal" href="#prepare-aggregated-junction-file">prepare aggregated junction file</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="install.html" title="previous chapter">Installation</a></li>
      <li>Next: <a href="modules.html" title="next chapter">jGEM packages</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/quickstart.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2016, Ken Sugino.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.4.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.7</a>
      
      |
      <a href="_sources/quickstart.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>